ssl/quic/quic_channel.c

```diff
@@ -0,0 +1,1661 @@
+{
+    return ch->handshake_complete;
+}
+
+/*
+ * QUIC Channel: Callbacks from Miscellaneous Subsidiary Components
+ * ================================================================
+ */
+
+/* Used by various components. */
+static OSSL_TIME get_time(void *arg)
+{
+    return ossl_time_now();
+}
+
+/* Used by QSM. */
+static uint64_t get_stream_limit(int uni, void *arg)
+{
+    QUIC_CHANNEL *ch = arg;
+
+    return uni ? ch->max_local_streams_uni : ch->max_local_streams_bidi;
+}
+
+/*
+ * Called by QRX to determine if a packet is potentially invalid before trying
+ * to decrypt it.
+ */
+static int rx_early_validate(QUIC_PN pn, int pn_space, void *arg)
+{
+    QUIC_CHANNEL *ch = arg;
+
+    /* Potential duplicates should not be processed. */
+    if (!ossl_ackm_is_rx_pn_processable(ch->ackm, pn, pn_space))
+        return 0;
+
+    return 1;
+}
+
+/*
+ * QUIC Channel: Handshake Layer Event Handling
+ * ============================================
+ */
+static int ch_on_crypto_send(const unsigned char *buf, size_t buf_len,
+                             size_t *consumed, void *arg)
+{
+    int ret;
+    QUIC_CHANNEL *ch = arg;
+    uint32_t enc_level = ch->tx_enc_level;
+    uint32_t pn_space = ossl_quic_enc_level_to_pn_space(enc_level);
+    QUIC_SSTREAM *sstream = ch->crypto_send[pn_space];
+
+    if (!ossl_assert(sstream != NULL))
+        return 0;
+
+    ret = ossl_quic_sstream_append(sstream, buf, buf_len, consumed);
+    return ret;
+}
+
+static int crypto_ensure_empty(QUIC_RSTREAM *rstream)
+{
+    size_t avail = 0;
+    int is_fin = 0;
+
+    if (rstream == NULL)
+        return 1;
+
+    if (!ossl_quic_rstream_available(rstream, &avail, &is_fin))
+        return 0;
+
+    return avail == 0;
+}
+
+static int ch_on_crypto_recv(unsigned char *buf, size_t buf_len,
+                             size_t *bytes_read, void *arg)
+{
+    QUIC_CHANNEL *ch = arg;
+    QUIC_RSTREAM *rstream;
+    int is_fin = 0; /* crypto stream is never finished, so we don't use this */
+    uint32_t i;
+
+    /*
+     * After we move to a later EL we must not allow our peer to send any new
+     * bytes in the crypto stream on a previous EL. Retransmissions of old bytes
+     * are allowed.
+     *
+     * In practice we will only move to a new EL when we have consumed all bytes
+     * which should be sent on the crypto stream at a previous EL. For example,
+     * the Handshake EL should not be provisioned until we have completely
+     * consumed a TLS 1.3 ServerHello. Thus when we provision an EL the output
+     * of ossl_quic_rstream_available() should be 0 for all lower ELs. Thus if a
+     * given EL is available we simply ensure we have not received any further
+     * bytes at a lower EL.
+     */
+    for (i = QUIC_ENC_LEVEL_INITIAL; i < ch->rx_enc_level; ++i)
+        if (i != QUIC_ENC_LEVEL_0RTT &&
+            !crypto_ensure_empty(ch->crypto_recv[ossl_quic_enc_level_to_pn_space(i)])) {
```

Comment:
Isn't all this checked in `ch_on_handshake_yield_secret`? Is it OTT to check it every time we receive crypto data too?

Reply:
We could still receive data on old ELs after `yield_secret`, no?

Reply:
Hmm...yes - good point. But wouldn't that be better handled on receipt in quic_rx_depact.c/depack_do_frame_crypto? i.e. confirm that either the received enc level >= rx_enc_level or, if not, that the stream remains empty after queuing the data?

Reply:
The QRX models this in terms of dropping ELs. We drop an EL and destroy its cryptographic material when we no longer want to receive packets for that EL. When to drop an EL is largely specified by the RFCs, however. We may not want to receive more CRYPTO frame data on an old EL but this doesn't mean the EL should necessarily be discarded yet.